### PR TITLE
Tpetra: multithread local matrix transpose

### PR DESCRIPTION
**Pull request by aprokop**
@trilinos/tpetra @mhoemmen 

First draft of local transpose multithreading.

Related to #2236.

Things I'm not sure about (and have not thought much):
1. While the class is called `RowMatrixTranposer`, it really only works for crs matrices. Should the row matrix path in local transpose be removed?
2. I'm not sure about using all those atomics.
3. I assume it would be nice to do 2-level parallelism.
4. The tests pass but I'm not sure if I'm running them right. I did add OMP_NUM_THREADS to environment when running, but maybe I need to do more.

### CODE REVIEW

File: packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp [hunk 1]
```diff
@@ -120,62 +120,131 @@ createTransposeLocal (const Teuchos::RCP<Teuchos::ParameterList> &params)
...
+        auto rowView = lclGraph.rowConst(row);
+        auto length  = rowView.length;
+
+        for (decltype(length) colID = 0; colID < length; colID++) {
+          auto col = rowView(colID);
+          Kokkos::atomic_fetch_add(&t_rows[col], 1);
+        }
+    });
+
+    // Compute offsets
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp]: Tpetra has a function for computing offsets from counts: `Tpetra::Details::computeOffsetsFromCounts` I think.  This should be fine, though.

File: packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp [hunk 2]
```diff
@@ -120,62 +120,131 @@ createTransposeLocal (const Teuchos::RCP<Teuchos::ParameterList> &params)
...
+    });
+
+    local_matrix_type lclTransposeMatrix("transpose", numLocalCols, numLocalRows, numLocalNnz, t_vals, t_rows, t_cols);
+
+    transMatrixWithSharedRows =
+        rcp (new crs_matrix_type (lclTransposeMatrix,
+                                  origMatrix_->getColMap (), origMatrix_->getRowMap (),
+                                  origMatrix_->getRangeMap (), origMatrix_->getDomainMap ()));
+  } else {
+    // FIXME: are we ever here? There is no RowMatrix constructor, we seem to
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp]: Huh, this is a good point.  It's not really a "RowMatrixTransposer" then ;-) .  It's up to you whether you would like to purge this code path.

Comment (aprokop) [reply packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp]: It seems that the proper solution would instead be to introduce a `RowMatrix` constructor, and store the row matrix that way. This way, it expands the applicability of the transposer. However, it's not clear to me if that is tested anywhere.

**Comment by aprokop** [comment]
@mhoemmen Can you address the atomics and 2-level parallelism question?

**Comment by aprokop** [comment]
Hmm, it seems like there is already some kind of transpose in [kokkos-kernels](https://github.com/kokkos/kokkos-kernels/blob/1a7b524ba38fdfab6c1058065af06cbcb4a2ce6f/src/common/KokkosKernels_SparseUtils.hpp).

**Comment by mhoemmen** [comment]
@aprokop wrote:
> I'm not sure about using all those atomics.

  - It's no different than the approach we've been taking for constructing the local sparse graph in finite-element assembly.
  - Another option would be to convert to coordinate (i,j) storage, flip pairs to (j,i), sort in reverse lexicographic order, and gather back into compressed sparse row storage.  However, that would take more memory, and the "gather back" step would require a parallel segmented scan implementation.
  - The common case has no more entries in a column than entries in a row, so the atomic updates during the "count" phase should not be heavily contended.

**Comment by aprokop** [comment]
@mhoemmen So, would you be OK with merging this and opening an issue to a) clean up RowMatrix stuff, and b) possibly replacing with kokkos kernel transpose operator, and c) possibly introducing dynamic scheduling?

**Comment by mhoemmen** [comment]
@aprokop wrote:
> So, would you be OK with merging this and opening an issue to a) clean up RowMatrix stuff, and b) possibly replacing with kokkos kernel transpose operator, and c) possibly introducing dynamic scheduling?

I just realized that this threaded version may have nondeterministic results, since it doesn't sort rows.  The current RowMatrixTransposer implementation doesn't sort, either.  Would it make sense to change current behavior by sorting rows?  That would make this more consistent with the usual `Crs{Graph,Matrix}::fillComplete`.

**Comment by aprokop** [comment]
@mhoemmen Damn, you are right. I really need to run experiments on Cori right now with this patch and hope it will work. I think the current serial version sorts, at least in terms of local indices, as it goes row by row. From MueLu's perspective, as long as matrix-matrix multiplication does not care about sorted indices, we don't care.

**Comment by mhoemmen** [comment]
@aprokop Ifpack2::LocalFilter assumes that each row has locals first, followed by remotes.

**Comment by mhoemmen** [comment]
@aprokop Many years ago, I tried fixing Ifpack2::LocalFilter so that it uses global indices to decide what's local vs. what's remote.  A bunch of tests broke.  It's likely because AdditiveSchwarz's overlap setup makes a bunch of weird assumptions that nothing else does.

My guess is that if the domain Map is ["locally fitted to"](https://github.com/trilinos/Trilinos/blob/69f8850c25bdead51b38866ae34d4fa40947b67f/packages/tpetra/core/src/Tpetra_Map_decl.hpp#L950) the column Map, and if the remotes only ever appear at the end of each row, then Ifpack2::LocalFilter should work just fine.  It shouldn't care otherwise about order.

The problem is that this issue may only manifest with some preconditioners.

**Comment by mhoemmen** [comment]
@aprokop This LocalFilter issue is old Bugzilla Bug 5992: https://software.sandia.gov/bugzilla/show_bug.cgi?id=5992 (need an account to log in).

**Comment by aprokop** [comment]
@mhoemmen Thanks! There is a lot of interesting tidbits there. But judging by the fact you were not successful, means that as long as we sort local indices, we have a chance to satisfy `LocalFilter`, right? Is there already a utility in Tpetra that does that?

**Comment by mhoemmen** [comment]
@aprokop wrote:
> But judging by the fact you were not successful, means that as long as we sort local indices, we have a chance to satisfy LocalFilter, right? 

> Is there already a utility in Tpetra that does that?

Not fully Kokkos-ized, alas.  Tpetra has some device functions that we could use to sort each row.  See e.g., `Tpetra_Details_{short,radix}Sort.hpp`.  Some Tpetra paths do the sort-and-merge in thread-parallel fashion on host.

**Comment by mhoemmen** [comment]
@aprokop I'm OK with merging this, though I'll mark it "Results Impacting" just in case.  All the sparse matrix-matrix multiply work happening recently is "results impacting" in the sense that it uses different algorithms.

**Comment by aprokop** [comment]
Thanks @mhoemmen!

**Pull request merged**